workflow/applications/applications.py

```diff
@@ -91,45 +83,49 @@ def __init__(self, conf: Configuration) -> None:
         self.aero_anl_runs = None
         self.aero_fcst_runs = None
         if self.do_aero:
-            aero_anl_run = _base.get('AERO_ANL_RUN', 'BOTH').lower()
+            aero_anl_run = self._base.get('AERO_ANL_RUN', 'BOTH').lower()
             if aero_anl_run in ['both']:
                 self.aero_anl_runs = ['gfs', 'gdas']
             elif aero_anl_run in ['gfs', 'gdas']:
                 self.aero_anl_runs = [aero_anl_run]
-            aero_fcst_run = _base.get('AERO_FCST_RUN', None).lower()
+            aero_fcst_run = self._base.get('AERO_FCST_RUN', None).lower()
             if aero_fcst_run in ['both']:
                 self.aero_fcst_runs = ['gfs', 'gdas']
             elif aero_fcst_run in ['gfs', 'gdas']:
                 self.aero_fcst_runs = [aero_fcst_run]
 
-    def _init_finalize(self, *args):
+    def _init_finalize(self, conf: Configuration):
         print("Finalizing initialize")
 
         # Get a list of all possible config_files that would be part of the application
         self.configs_names = self._get_app_configs()
 
-        # Source the config_files for the jobs in the application
-        self.configs = self.source_configs()
+        # Source the config files for the jobs in the application without specifying a RUN
+        self.configs = {'_no_run': self._source_configs(conf)}
```

Comment:
This is because of the subclass `update_base` call in `init_finalize`.  Currently, this method is only called for the `_no_run` configuration dictionary, so only that `config['base']` has, e.g., `SDATE_GFS` in it.  I could rework this so that `update_base` is called for all configuration dictionaries.  Would you prefer that?

If we go this route, and we want to specify a `run` for all downstream references to the configuration dictionaries, then we will need to know which `run` to call at the appropriate time.  For instance, a cycled experiment may or may not have a `gfs` `run` while a forecast-only experiment will only have a `gfs` or `gefs` run.  We could solve this by retrieving an arbitrary `run`, i.e.
```
arb_run=next(iter(configs.values()))
```

But if we go that route, then a `_no_run` configuration dictionary looks more attractive as you will always end up with the same `config.base`.  But if we plan on adding `RUN`-specific variables to `config.base` that are needed at setup time, then both of these approaches may not suffice long term.

The last option is to abandon `_no_run` and add a function with logic to check for available `run`s and pick the most logical choice.  E.g.
```
def select_run(configs, mode)
  cycled = mode == "cycled"
  if cycled:
    if "gfs" in configs:
      return "gfs"
   elif "gefs" in configs: # Assuming gefs will one day cycle
      return "gefs"
   elif "gdas" in configs:
     return "gdas"
   else:
     raise Exception("Invalid configuration setup!")
  else:
    if "gfs" in configs:
      return "gfs"
    elif "gefs" in configs":
      return "gefs"
    else:
      raise Exception("Invalid configuration setup!")
```

Is this last option what you would like to see?  Or is there more down the road that we need to consider before selecting another option?

Reply:
The last option is closer to what I was thinking, but I think we need to do more here. We're already building a list of component runs in init(), but even that I think should be changed so that `do_*` is run-based. This lets us cleanly just change the APP based on RUN in `config.base` and have everything that follows from that just work. This might require moving them to properties of Task instead of AppConfig.

Reply:
I wasn't able to implement the last option completely.  Instead, I isolated the `_no_run` to just the `_init_finalize` method where the valid `run`s are determined.  After that, `run` is used when referencing any config.

I started to look into moving the `do_*` flags to `Tasks`.  Since `_get_app_configs` needs to know these flags when `_init_finalize` is called in order to determine which configs to include, I don't think it is feasible to move them.  However, I do think they can be set based on `RUN` and determined within `AppConfig` subclasses.  I'll look into this more next week.